### ISSUE TITLE
Conditional Logic / Shuffling

### ISSUE THREAD

**Issue by NickBanana7** [GioanZ/Portfolio-Stress-Testing-with-Deep-Generative-Models#1]
Hi! Thanks for sharing this repo - nice to see Conditional VAE and GAN combined - where such setups would benefit from nonlinear dimensionality reduction (VAE) via latent space and the ability of the graph network to learn complex data patterns (GAN). Also, it is nice that you have considered custom loss functions, taking into account distribution quantiles (worst return outcomes).

I've just had a (few) comments and questions regarding the implementation logic of some parts, please see below: 

- in the data preparation part, two datasets consisting of input data (returns) and label or conditioning data (macro) are processed and aligned
- after performing z-score transformation on both datasets, they are passed to the CVAE model. 
- next, after the CVAE training is complete, the encoder is involed and mean latent output is passed as input to the CGAN model

When setting up the CVAE architecture shouldn't the condition dimension be concatenated with both encoder and decoder models? In your case, the encoder model is sampling unconditional noise and then in the decoder the sampled z output is concatenated with condtional input of shape _macro_dim._

During model training, the entire labels/macro dataset is passed to the CVAE/CGAN models. Here did you consider enforcing a conditional logic based on multiple past returns to forecast future returns?

For example:
- use _h_ past return samples to predict future _f_ returns (h,f)
- this logic would help capture serial correlation effects 
- align with business simulation logic and/or VAR calculation logic
- provide a richer Stress testing framework, as currently only 1 value of each column can be used for a stress scenario (kindly correct me if I am wrong here) - which perhaps does not realistically capture such events.

To implement such a logic, one would need to change the input dataset, for both returns and labels/macro inputs. See below a small function I did for these purposes, to be used in data preparation (after returns/macro transformation for example): 
```
#Create windows

def create_windows(data, nh, nf):
    X, Y = [], []
    for i in range(len(data) - nh - nf + 1): 
        # Input data: nh (historical days)
        X.append(data[i: i + nh])
        # Target data: nf (forecast days)
        Y.append(data[i + nh: i + nh + nf])
    
    # Convert to numpy arrays and output:
    return np.array(X), np.array(Y)
```

Of course, this could imply changes in both the model training and simulation sections of the code. 
Also, I've noticed that in the _train_conditional_wgan_gp_ function, when creating the _train_dataset_ variable a `.shuffle() ` method is being called. This is fine for cross-sectional data, but for time-series datasets shuffling the dataset prior to model building indicates that the model would not be learning from the correct sequence of events, breaking temporal dependencies and potentially leading to poor generalization. Here, returns usually do not display significant serial dependence, absolute and squared returns do (stylized features of financial time-series). However, shuffling before training would not be compatible with the above described conditional logic. 

Anyhow, in such a setup when passing a 3-D tensor [N-samples, T-timesteps, D-dimensions] to CVAE - also the logic of the reconstruction loss e.g., by feature and/or axis dimension should be updated.

Let me know your thoughts! Have a nice day.
